File: solution/0300-0399/0376.Wiggle Subsequence/Solution.go
func wiggleMaxLength(nums []int) int {
	f, g := 1, 1
	for i, x := range nums[1:] {
		d := x - nums[i]
		if d < 0 {
			f = max(f, g+1)
		}
		if d > 0 {
			g = max(g, f+1)
		}
	}
	return max(f, g)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: solution/0300-0399/0376.Wiggle Subsequence/Solution.java
class Solution {
    public int wiggleMaxLength(int[] nums) {
        int f = 1, g = 1;
        for (int i = 1; i < nums.length; ++i) {
            int d = nums[i] - nums[i - 1];
            if (d < 0) {
                f = Math.max(f, g + 1);
            }
            if (d > 0) {
                g = Math.max(g, f + 1);
            }
        }
        return Math.max(f, g);
    }
}